plugins/inputs/fritzbox/fritzbox.go

```diff
@@ -0,0 +1,471 @@
+	WanInfo           bool            `toml:"wan_info"`
+	PppInfo           bool            `toml:"ppp_info"`
+	DslInfo           bool            `toml:"dsl_info"`
+	WlanInfo          bool            `toml:"wlan_info"`
+	HostsInfo         bool            `toml:"hosts_info"`
+	FullQueryCycle    int             `toml:"full_query_cycle"`
+	Timeout           config.Duration `toml:"timeout"`
+	TlsSkipVerify     bool            `toml:"tls_skip_verify"`
+	Debug             bool            `toml:"debug"`
+	Log               telegraf.Logger `toml:"-"`
+	deviceClients     []*tr064.Client
+	serviceHandlers   map[string]serviceHandlerFunc
+	queryCycleCounter int
+}
+
+func defaultFritzbox() *Fritzbox {
+	return &Fritzbox{
+		Devices:           make([]string, 0),
+		DeviceInfo:        true,
+		WanInfo:           true,
+		PppInfo:           true,
+		DslInfo:           true,
+		WlanInfo:          true,
+		HostsInfo:         false,
+		FullQueryCycle:    30,
+		Timeout:           defaultTimeout,
+		deviceClients:     make([]*tr064.Client, 0),
+		serviceHandlers:   make(map[string]serviceHandlerFunc),
+		queryCycleCounter: 0,
+	}
+}
+
+func (*Fritzbox) SampleConfig() string {
+	return sampleConfig
+}
+
+func (plugin *Fritzbox) Init() error {
+	if plugin.Log == nil {
+		plugin.Log = logger.New("inputs", pluginName, "")
+	}
+	err := plugin.initDeviceClients()
+	if err != nil {
+		return err
+	}
+	plugin.initServiceHandlers()
+	return nil
+}
+
+func (plugin *Fritzbox) initDeviceClients() error {
+	for _, device := range plugin.Devices {
+		deviceUrl, err := url.Parse(device)
+		if err != nil {
+			return err
+		}
+		client := tr064.NewClient(deviceUrl)
+		client.Debug = plugin.Debug
+		client.Timeout = time.Duration(plugin.Timeout)
+		client.InsecureSkipVerify = plugin.TlsSkipVerify
+		plugin.deviceClients = append(plugin.deviceClients, client)
+	}
+	return nil
+}
+
+func (plugin *Fritzbox) initServiceHandlers() {
+	if plugin.DeviceInfo {
+		plugin.serviceHandlers[deviceinfo.ServiceShortType] = plugin.gatherDeviceInfo
+	}
+	if plugin.WanInfo {
+		plugin.serviceHandlers[wancommonifconfig.ServiceShortType] = plugin.gatherWanInfo
+	}
+	if plugin.PppInfo {
+		plugin.serviceHandlers[wanpppconn.ServiceShortType] = plugin.gatherPppInfo
+	}
+	if plugin.DslInfo {
+		plugin.serviceHandlers[wandslifconfig.ServiceShortType] = plugin.gatherDslInfo
+	}
+	if plugin.WlanInfo {
+		plugin.serviceHandlers[wlanconfig.ServiceShortType] = plugin.gatherWlanInfo
+	}
+	if plugin.HostsInfo {
+		plugin.serviceHandlers[hosts.ServiceShortType] = plugin.gatherHostsInfo
+	}
+}
+
+func (plugin *Fritzbox) Gather(acc telegraf.Accumulator) error {
+	start := time.Now()
+	for _, deviceClient := range plugin.deviceClients {
+		plugin.gatherDevice(acc, deviceClient)
+	}
+	if plugin.Debug {
+		elapsed := time.Since(start)
+		plugin.Log.Info("took: ", elapsed)
+	}
+	plugin.queryCycleCounter++
+	if plugin.queryCycleCounter >= plugin.FullQueryCycle {
+		plugin.queryCycleCounter = 0
+	}
```

Comment:
This can be replaced by using the `inputs.internal` metrics.

Reply:
Will remove

Reply:
Done via ccd459aee13ea59a56e50afcc9f0e590f6967638